test/server/media-controller.spec.js

```diff
@@ -0,0 +1,150 @@
+const chai = require( 'chai' );
+const expect = chai.expect;
+const request = require( 'request' );
+const config = require( '../../app/models/config-model' ).server;
+
+const IPfiltering = config[ 'ip filtering' ];
+
+/**
+ * Tests the request-filtering-agent to block SSRF attacks
+ * change testHTMLBody to the body of an html file that
+ * you are testing on. For the default, it says <i'm in.>
+ * and is hosted in testHTMLHost. 
+ */
+
+const testHTMLBody = 'i\'m in.\n';
+const enketoHost = 'http://localhost:8005';
+const testHTMLHost = 'http/localhost:1234';
+
+const requestURL = enketoHost + '/media/get/' + testHTMLHost;
+
+// TODO: Check if testHTMLHost is running
+
+describe( 'Media Controller', () => {
+
+    console.log( 'Testing request-filtering-agent on this request: ' + requestURL);
+    
+    describe( 'No Referer Request', () => {
+
+        if (!IPfiltering.allowPrivateIPAddress) {
```

Comment:
What would be very valuable is if we can test all the cases in the if statements on this page (so 12 instead of 4). Did you try this? I see we may be doing something similar in api-controller.spec.js (with `app.set`).

We want to be very comfortable that we always test the main cases, to be confident that any future library updates don't break anything.

Reply:
Not sure exactly what you mean here, but I've remade the test suite to be a bit more readable and (hopefully) running properly this time.

Reply:
What I meant was in your if() conditions there are 3 tests but only 1 is actually tested, depending on the configuration (config.json). We need to test all 3, by dynamically updating the configuration.

Reply:
@duvld, just checking not pressuring, will you still update the 12 tests, and remove all if() statements?

Reply:
Yep! Hopefully I'll get to it today or tomorrow :)

Reply:
@MartijnR , I've tried doing the following for dynamically updating the configuration:
```
const app = require( '../../config/express' );
const filteringOptions = app.get('ip filtering');
filteringOptions['allowPrivateIPAddress'] = true;
filteringOptions['allowMetaIPAddress'] = false;
filteringOptions['allowIPAddressList'] = [];
filteringOptions['denyIPAddressList'] = [];
app.set('ip filtering', filteringOptions);
```
This works for setting the configuration correctly, but somehow my tests no longer work - even though this shouldn't be any different than me manually updating my `config.json`. This looks like it'll take me a bit longer than expected, but I wanted to ask in case you see something that I'm not.

Reply:
Am checking!

Reply:
Could you push that commit(s) please?

Reply:
Just a question, in case that is it: Are you making the request like this: `request( app ).get(...)`

Reply:
Sorry for the late response! I pushed my changes here: https://github.com/enketo/enketo-express/pull/150/commits/1ad19b2154fa612cf4955ecef9212d71973adf0c
I also found that when I call `app.set` in the test it doesn't change in `media-controller.js`, there the options still are identical to the ones in `config.json`. I've tried to use `app.get/set` in `media-controller.js` but it `get` returns `{}` (so `set` doesn't work).

Reply:
I noticed the same issue (eventually) and realized, we needed to change the media-controller just to make it testable. See the commit I added. 836eae3dc0c2f507be4a45dce2ab83f1cbf839ad

I was initially working on changing the tests themselves, and have added those changes as a separate commit in case they are useful: 51ff99d2c5cbab735c5c3553919f3eaf7e0042a6 Feel free to revert.

Reply:
Wow, I was trying different stuff all day to get `media-controller` to take on the changes. If you can't tell I'm new to this side of javascript :bow: . I'll definitely take a look at these tests :) Thanks!

Reply:
Hi @MartijnR , the tests for referred requests are looking good and completed. I've pushed them up here: https://github.com/enketo/enketo-express/pull/150/commits/f1ebab01b5f280ac1c6c00df94412c30896c333d. Now I'm running into a weird problem with non referred requests. I think it has something to do with how I'm using `supertest`. I removed the `.set(...)` from the list of calls on `request(...)`:
```
    it( 'for a private IP address WITH a Referer with allowPrivateIPAddress=false', done => {

        // Don't change any default IP filtering setting
        app.set( 'ip filtering', { allowPrivateIPAddress, allowMetaIPAddress, allowIPAddressList, denyIPAddressList } );

        request( app )
            .get( requestURL )
            .expect( 500, /DNS lookup .* is not allowed. Because, It is private IP address/)
            .end( done );
    } );
```
This is giving me the expected errors, but mocha is telling me it's "Uncaught":
```
1) Testing request-filtering-agent
       for a private IP address WITHOUT a Referer with allowPrivateIPAddress=false:
     Uncaught Error: DNS lookup 127.0.0.1(family:4, host:localhost) is not allowed. Because, It is private IP address.
      at validateIPAddress (node_modules/request-filtering-agent/lib/request-filtering-agent.js:58:20)
      at Socket.<anonymous> (node_modules/request-filtering-agent/lib/request-filtering-agent.js:75:21)
      at GetAddrInfoReqWrap.emitLookup [as callback] (net.js:1044:12)
      at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:66:8)
```
I've tried making try catch statements and that hasn't worked either

Reply:
I'll have a look! 

Btw, the previous issue, made me realize it is probably never a good idea to require the config-model as I had suggested earlier. I should probably go through the code remove those instances. Sorry about that.

Reply:
I think I found it! The way I saw it was that the non-referred requests were being handled in `media-controller` made throwing the error occur after the get request finishes. I've changed around `media-controller` a bit here: https://github.com/enketo/enketo-express/pull/150/commits/78c95abc31aebf2476209a543821c8efe909fec7, and all the tests seem to work now. Side note, I'm almost 100% sure the way I changed `media-controller` is not the _best_ way to do it.

Reply:
That's progress! I think this duplicates all allowed media requests though.

Would putting request.get inside a try..catch statement work perhaps?

Reply:
This seems to do the trick: 
```
function _pipeMedia( options, req, res, next ) {
    request.get( options ).on( 'error', error => _handleMediaRequestError( error, next ) ).pipe( res ).on( 'error', error => _handleMediaRequestError( error, next ) );
}

function _handleMediaRequestError( error, next ) {
    debug( `error retrieving media from OpenRosa server: ${JSON.stringify( error )}` );
    if ( !error.status ) {
        error.status = ( error.code && error.code === 'ENOTFOUND' ) ? 404 : 500;
    }
    next( error );
}
```

Reply:
Please verify and if okay commit, then I think we're good to squash and merge. (I don't want to add the last commit myself to keep you as commit author after squashing).

Reply:
Never mind, this seems to break obtaining media. :)

Reply:
Actually, it's caused by the IP filtering module. I'm seeing the error: TypeError [ERR_INVALID_PROTOCOL]: Protocol "https:" not supported. Expected "http:"

Reply:
closing this as I prepared a new review that includes these remaining comments